### ISSUE TITLE
Create Artifact abstraction class

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#908]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/908)
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 12, 2019, 15:07

## Background

For [Artifact as a Proto](https://mail.gnome.org/archives/buildstream-list/2019-February/msg00027.html) an abstract Artifact class, which will implement artifact specific functions which currently reside in Element.

## Task description

* Create list of existing functions that should be abstracted out of Element via Artifact

Current list of potential artifact related funcs to abstract:

* [x]   __compute_splits (imports files dir of artifact)
* [x]   stage_artifact (imports files dir of artifact)
* [x]   _stage_sources_at (checks if a requested buildtree is cached empty)
* [x]   _cache_artifact
* [x]   _get_build_log
* [x]   _cached_buildtree
* [x]   __extract
* [x]   __get_artifact_metadata_keys
* [x]   __get_artifact_metadata_dependencies
* [x]   __get_artifact_metadata_workspaced
* [x]   __get_artifact_metadata_workspaced_dependencies
* [x]   __load_public_data
* [x]   __load_build_result

* Abstract out, no new tests to be added

## Acceptance Criteria

Current bst functionality is preserved

----

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 12, 2019, 15:09

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 12, 2019, 15:09

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 13, 2019, 15:08

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 13, 2019, 15:08

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 14, 2019, 13:09

marked this issue as related to #909

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 14, 2019, 16:11

I expect to prune the list once implementation starts. Creating the class and then moving methods over piece by piece, ensuring tests work between each commit. Core abstractions should be in place before 'nice to have' refactors are considered.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 14, 2019, 16:52

*Note abstraction not used in the strict sense of abstract base classes

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 18, 2019, 15:06

I plan to move over aspects of ` __load_build_result` & `_cached_buildtree` as initial abstractions

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 10:48

marked the task **__extract** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 10:48

marked the task **__load_build_result** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 10:49

marked the task **__load_public_data** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 10:49

marked the task **_cache_artifact** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 10:49

marked the task **_cached_buildtree** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 11:09

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 14:17

marked the task **__get_artifact_metadata_keys** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 14:17

marked the task **__get_artifact_metadata_dependencies** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 16:02

marked the task **__get_artifact_metadata_workspaced** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 16:02

marked the task **__get_artifact_metadata_workspaced_dependencies** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 16:47

marked the task **_get_build_log** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 21, 2019, 16:51

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 22, 2019, 14:57

mentioned in merge request !1175

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 22, 2019, 14:58

I've pushed a WIP for this with my current transitions

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 26, 2019, 11:33

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 26, 2019, 14:26

marked the task **_stage_sources_at (checks if a requested buildtree is cached empty)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 26, 2019, 15:20

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 26, 2019, 15:22

marked the task **stage_artifact (imports files dir of artifact)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 26, 2019, 15:22

marked the task **__compute_splits (imports files dir of artifact)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Mar 5, 2019, 15:25

marked this issue as related to #945

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @marge-bot123]](https://gitlab.com/marge-bot123) on Mar 12, 2019, 21:16

closed via merge request !1175

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @marge-bot123]](https://gitlab.com/marge-bot123) on Mar 12, 2019, 21:16

mentioned in commit 2c3dfa4e1b92ead066b3fa25fe7c3037b4ba6c58
